dagger/manifest.nim

```diff
@@ -0,0 +1,197 @@
+import pkg/questionable
+import pkg/questionable/results
+import pkg/chronicles
+import pkg/chronos
+
+import ./blocktype
+
+const
+  ManifestCodec* = multiCodec("dag-pb")
+
+type
+  BlocksManifest* = object
+    blocks: seq[Cid]
+    htree: ?Cid
+    version*: CidVersion
+    hcodec*: MultiCodec
+    codec*: MultiCodec
+    cidEmptyDigest: MultiHash
+
+proc len*(b: BlocksManifest): int = b.blocks.len
+
+iterator items*(b: BlocksManifest): Cid =
+  for b in b.blocks:
+    yield b
+
+proc hashBytes(mh: MultiHash): seq[byte] =
+  mh.data.buffer[mh.dpos..(mh.dpos + mh.size - 1)]
+
+proc cid*(b: var BlocksManifest): ?Cid =
+  if b.htree.isSome:
+    return b.htree
+
+  var
+    stack: seq[MultiHash]
+
+  if stack.len == 1:
+    stack.add(b.cidEmptyDigest)
+
+  for cid in b.blocks:
+    if mhash =? cid.mhash:
+      stack.add(mhash)
+
+    while stack.len > 1:
+      var
+        (b1, b2) = (stack.pop(), stack.pop())
+
+      var
+        digest = MultiHash.digest(
+          $b.hcodec,
+          (b1.hashBytes() & b2.hashBytes()))
+
+      without mh =? digest:
+        return Cid.none
+
+      stack.add(mh)
+
+  if stack.len == 1:
+    let
+      cid = Cid.init(b.version, b.codec, stack[0])
+
+    if cid.isOk:
+      b.htree = cid.get().some
+      return cid.get().some
+
+proc put*(b: var BlocksManifest, cid: Cid) =
+  if b.htree.isSome:
+    b.htree = Cid.none
+
+  trace "Adding cid to manifest", cid
+  b.blocks.add(cid)
+
+proc contains*(b: BlocksManifest, cid: Cid): bool =
+  cid in b.blocks
+
+proc encode*(b: var BlocksManifest): ?!seq[byte] =
+  ## Encode the manifest into a ``ManifestCodec``
+  ## multicodec container (Dag-pb) for now
+  var pbNode = initProtoBuffer()
+
+  for c in b.blocks:
+    var pbLink = initProtoBuffer()
+    pbLink.write(1, c.data.buffer) # write Cid links
+    pbLink.finish()
+    pbNode.write(2, pbLink)
+
+  without cid =? b.cid:
+    return failure("Unable to generate tree hash")
+
+  pbNode.write(1, cid.data.buffer) # set the treeHash Cid as the data field
+  pbNode.finish()
+
+  return pbNode.buffer.success
+
+proc decode*(_: type BlocksManifest, data: seq[byte]): ?!(Cid, seq[Cid]) =
+  ## Decode a manifest from a byte seq
+  ##
+  var pbNode = initProtoBuffer(data)
+  var
+    cidBuf: seq[byte]
+    blocks: seq[Cid]
+
+  if pbNode.getField(1, cidBuf).isOk:
+    if cid =? Cid.init(cidBuf):
+      var linksBuf: seq[seq[byte]]
+      if pbNode.getRepeatedField(2, linksBuf).isOk:
+        for pbLinkBuf in linksBuf:
+          var
+            blocksBuf: seq[seq[byte]]
+            blockBuf: seq[byte]
+            pbLink = initProtoBuffer(pbLinkBuf)
+
+          if pbLink.getField(1, blockBuf).isOk:
+            let cidRes = Cid.init(blockBuf)
+            if cidRes.isOk:
+              blocks.add(cidRes.get())
+
+      return (cid, blocks).success
+
+func init*(
+  T: type BlocksManifest,
+  blocks: openArray[Cid] = [],
+  version = CIDv1,
+  hcodec = multiCodec("sha2-256"),
+  codec = multiCodec("raw")): ?!T =
```

Comment:
The reason we use `CatchableError` is because it's a lot easier to map across different error types and it avoids explicit mapping from one error type to another. It's also just as inspectable as an `enum`, because you can check if the instance is of some specific type with the `is` operator:

```nim
type
  DaggerError                  = object of CatchableError
  NotDaggerError               = object of CatchableError
  CreateEmptyCidFailure        = object of DaggerError

block:
  echo "\n***CreateEmptyCidFailure test***"
  let exc = (ref CreateEmptyCidFailure)(msg: "Unable to get multihash from empty Cid")

  if exc is ref CreateEmptyCidFailure:
    echo exc.msg
  elif exc is ref DaggerError:
    echo "Generic error!"
  elif exc is ref NotDaggerError:
    echo "Should not hit"
  else:
    doAssert false

block:
  echo "\n***DaggerError test***"
  let exc = (ref DaggerError)()

  if exc is ref CreateEmptyCidFailure:
    echo exc.msg
  elif exc is ref DaggerError:
    echo "Generic error!"
  elif exc is ref NotDaggerError:
    echo "Should not hit"
  else:
    doAssert false

block:
  echo "\n***Not DaggerError test***"
  let exc = (ref NotDaggerError)()

  if exc is ref CreateEmptyCidFailure:
    echo exc.msg
  elif exc is ref DaggerError:
    echo "Generic error!"
  elif exc is ref NotDaggerError:
    echo "Not a dagger error!"
  else:
    doAssert false

block:
  echo "\n***Catch All test***"
  let exc = (ref CatchableError)()

  if exc is ref CreateEmptyCidFailure:
    echo exc.msg
  elif exc is ref DaggerError:
    echo "Generic error!"
  elif exc is ref NotDaggerError:
    echo "Not a dagger error!"
  elif exc is ref CatchableError:
    echo "Catch All!"
  else:
    doAssert false
```

If you run the code, it will hit all the proper types depending on the one that got instantiated, but it also allows you to have more generic error handling logic that would otherwise be harder to implement.

Reply:
Thanks for all the info. I think this has been resolved now and to clarify iirc, the issue was that everything was `CatchableError` or `DaggerError` but no exceptions were derived past `DaggerError` meaning that it would have been difficult for caller to do what you have illustrated above.